pkg/registry/populator_test.go

```diff
@@ -1534,3 +1534,759 @@ func TestSemverPackageManifest(t *testing.T) {
 		})
 	}
 }
+
+func TestSubstitutesFor(t *testing.T) {
+	type args struct {
+		bundles  []string
+		rebuilds []string
+	}
+	type replaces map[string]string
+	type expected struct {
+		bundles        []string
+		channels       []string
+		substitutions  map[string]string
+		whatReplaces   map[string]replaces
+		channelHeads   map[string]string
+		defaultChannel string
+	}
+	tests := []struct {
+		description string
+		args        args
+		expected    expected
+	}{
+		{
+			description: "SubstitutesFor/FirstSubstitutionReplaces",
+			args: args{
+				bundles: []string{
+					"prometheus.0.22.2",
+					"prometheus.0.14.0",
+					"prometheus.0.15.0",
+				},
+				rebuilds: []string{
+					"prometheus.0.15.0.substitutesfor",
+				},
+			},
+			expected: expected{
+				bundles: []string{
+					"quay.io/test/prometheus.0.22.2",
+					"quay.io/test/prometheus.0.14.0",
+					"quay.io/test/prometheus.0.15.0.substitutesfor",
+					"quay.io/test/prometheus.0.15.0.substitutesfor",
+					"quay.io/test/prometheus.0.15.0",
+					"quay.io/test/prometheus.0.15.0",
+					"quay.io/test/prometheus.0.14.0",
+				},
+				channels: []string{
+					"stable",
+					"preview",
+					"preview",
+					"stable",
+					"stable",
+					"preview",
+					"preview",
+				},
+				substitutions: map[string]string{
+					"prometheusoperator.0.22.2":                      "",
+					"prometheusoperator.0.14.0":                      "",
+					"prometheusoperator.0.15.0":                      "",
+					"prometheusoperator.0.15.0+1-freshmaker-rebuild": "prometheusoperator.0.15.0",
+				},
+				whatReplaces: map[string]replaces{
+					"prometheusoperator.0.22.2": {
+						"preview": "",
+					},
```

Comment:
The fact that a "blank replaces" is needed to indicate channel membership makes me think we want to change around the relationship here. Is there a practical reason this is preferred over [the example in the previous comment](https://github.com/operator-framework/operator-registry/pull/612/files#r603506828)?

Reply:
I'm not sure I understand. Are you just saying that checking channel heads and channel membership is redundant after this check?

Reply:
I'm saying that `<bundle> -> []<channel>:<replaces>` is hard to grok and looks odd when you have a single node graph for a channel. I'm suggesting that `channel -> []<bundle>:<replacement>` conveys the same information without looking odd for that edge case (also, I think it gives a cleaner view of the actual path an upgrade could take).